prow/tide/codereview.go

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tide
+
+import (
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+type CodeReviewCommon struct {
+	// NameWithOwner is from graphql.NameWithOwner, <org>/<repo>
+	NameWithOwner string
+	// The number of PR
+	Number int
+	Org    string
+	Repo   string
+	// BaseRefPrefix gets prefix of ref, such as /refs/head, /refs/tags
+	BaseRefPrefix string
+	BaseRefName   string
+	HeadRefName   string
+	HeadRefOID    string
+
+	Title       string
+	Body        string
+	AuthorLogin string
+	// Labels gets labels on the PR.
+	// TODO(chaodaiG): labels might mean something different on gerrit, consider
+	// how to name this nicely.
+	Labels        []string
+	IsMergeable   bool
+	UpdatedAtTime time.Time
+
+	Mergeable    string
+	CanBeRebased bool
+	LogFields    logrus.Fields
+
+	// Likely GitHub only
+	Milestone      *Milestone
+	Commits        Commits
+	ReviewDecision string
+	RepoWithOwner  string
+}
+
+func (crc *CodeReviewCommon) logFields() logrus.Fields {
+	return logrus.Fields{
+		"org":    crc.Org,
+		"repo":   crc.Repo,
+		"pr":     crc.Number,
+		"branch": crc.BaseRefName,
+		"sha":    crc.HeadRefOID,
+	}
+}
+
+func CodeReviewCommonFromPullRequest(pr *PullRequest) *CodeReviewCommon {
```

Comment:
I think it would be good to test this function and to do it in a way where the test will fail if add new fileds and forget to update this. The best I can think of to make that happen would be to also build the reverse (`CodeReviewToPullRequest`) and then use fuzzed data and check if we get the same `PullRequest` after a roundtrip to `CodeReviewCommon` and back

Reply:
Totally agreed. I have done an experiment earlier with `PullRequest` and `gerrit.ChangeInfo` as substructs under `CommonCodeReview` struct, fuzzing `PullRequest` works as expected, but `gerrit.ChangeInfo` takes forever to fuzz and timed out after 30s, so we will run into trouble with gerrit with this roundtrip test. Will need to think a little more on how to make this work